src/iterators.rs

```diff
@@ -413,6 +415,50 @@ enum SliceOrInt<'a> {
     Int(isize),
 }
 
+trait PyConvertToPyArray {
+    fn convert_to_pyarray(&self, py: Python) -> PyResult<PyObject>;
+}
+
+macro_rules! py_convert_to_py_array_impl {
+    ($($t:ty)*) => ($(
+        impl PyConvertToPyArray for Vec<$t> {
+            fn convert_to_pyarray(&self, py: Python) -> PyResult<PyObject> {
+                Ok(self.clone().to_pyarray(py).into())
```

Comment:
Is there any downside of using `IntoPyArray` which doesn't allocate more memory? https://docs.rs/numpy/latest/numpy/array/struct.PyArray.html#memory-location

```suggestion
                Ok(self.clone().into_pyarray(py).into())
```

Reply:
We can not drop `clone()` if we use [`IntoPyArray`](https://docs.rs/numpy/latest/numpy/convert/trait.IntoPyArray.html) since it needs to take ownership of the vector but `__array__` receives a ref `&self` (and pyo3 doesn't allow to receive `self` and for a good reason since you can not move data from Python).

Reply:
Yeah, the underlying trait method `into_pyarray` (which is part of the `IntoPyArray` @georgios-ts linked to) is doing exactly that for us. It uses an unsafe block to get the pointer: https://github.com/PyO3/rust-numpy/blob/19bfc9d2d0e72bb3ed30c08244ee81abd02d7386/src/convert.rs#L67

Reply:
I think that's missing the lifetime/read-only ties, though, since it takes `self` not `&mut self`.

Reply:
@jakelishman Yeah, this is indeed possible and rust-numpy provides the unsafe [`borrow_from_array`](https://docs.rs/numpy/latest/numpy/array/struct.PyArray.html#method.borrow_from_array) (which results in a writeable array but you can't resize it) so we can avoid cloning the data.

Reply:
I switched to using `into_pyarray`